src/Repository/WikiRepository.php

```diff
@@ -151,6 +151,124 @@ public function getLivePagesWithWikiProjects( array $titlesByNs ): array {
+					WHERE ct_rev_id = revs.rev_id
+				 ) AS `tags`",
+			] )
+			->from( "{$this->getDb()}.revision", 'revs' )
+			->leftJoin( 'revs', "{$this->getDb()}.revision", 'parentrevs', 'revs.rev_parent_id = parentrevs.rev_id' )
+			->leftJoin( 'revs', "{$this->getDb()}.comment", 'comment', 'revs.rev_comment_id = comment.comment_id' )
+			->where( 'revs.rev_id IN (:revIds)' )
+			->setParameter( 'revIds', $revIds, ArrayParameterType::INTEGER );
+
+		$results = $qb->executeQuery()->fetchAllAssociative();
+
+		// Fetch localized tag labels.
+		$tagNames = [];
+		foreach ( $results as $result ) {
+			if ( isset( $result['tags'] ) ) {
+				$tagNames = array_merge( $tagNames, array_filter( explode( ',', $result['tags'] ) ) );
+			}
+		}
+		$tagLabels = $this->getTagsLabels( array_unique( $tagNames ) );
+
+		// Loop through again, adding in the tag labels.
+		foreach ( $results as $index => $result ) {
+			if ( isset( $result['tags'] ) ) {
+				$tagNames = array_filter( explode( ',', $result['tags'] ) );
+				$results[$index]['tags'] = array_filter( array_map( static function ( $tagName ) use ( $tagLabels ) {
+					return $tagLabels[$tagName];
+				}, $tagNames ) );
+			}
+		}
+
+		return $results;
+	}
+
+	/**
+	 * Get the labels for the given tags, as defined by their on-wiki interface messages.
+	 *
+	 * @param array $tagNames
+	 * @return array
+	 */
+	private function getTagsLabels( array $tagNames ): array {
+		$cacheKey = $this->getLang() . '-tags';
+
+		// Loop through to determine which messages we don't already have in cache.
+		$messages = [];
+		if ( $this->cache->hasItem( $cacheKey ) ) {
+			$messages = $this->cache->getItem( $cacheKey )->get();
+			foreach ( $tagNames as $index => $tagName ) {
+				if ( isset( $messages[$tagName] ) ) {
+					unset( $tagNames[$index] );
+				}
+			}
+		}
+
+		if ( !$tagNames ) {
+			return $messages;
+		}
+
+		$lang = $this->getLang();
+		$tagsToQuery = array_map( static function ( $tagName ) {
+			return "Tag-$tagName";
+		}, array_splice( $tagNames, 0, 50 ) );
+		$response = $this->httpClient->request(
+			'GET',
+			"https://$lang.wikipedia.org/w/api.php",
+			[
+				'query' => [
+					'action' => 'query',
+					'meta' => 'allmessages',
+					'ammessages' => implode( '|', $tagsToQuery ),
+					'amenableparser' => 1,
+					'amincludelocal' => 1,
+					'formatversion' => 2,
+					'format' => 'json',
+				]
+			]
+		)->toArray( false );
+
+		if ( $response ) {
+			$cacheItem = $this->cache->getItem( $cacheKey );
+			foreach ( $response['query']['allmessages'] as $message ) {
+				$tagName = preg_replace( '/^Tag\-/', '', $message['name'] );
+				$messages[$tagName] = $message['content'] ?? $tagName;
+				// Fill in blank values for hidden tags.
+				$messages[$tagName] = $messages[$tagName] === '-' ? '' : $messages[$tagName];
+				// Cache for a week.
+				$cacheItem->set( $messages )
+					->expiresAfter( new DateInterval( 'P7D' ) );
+				$this->cache->save( $cacheItem );
```

Comment:
This overwrites the cached item on every iteration of the loop. Is that intentional?

Reply:
No it is not! Thanks for catching this. Fixed :)